error-msg: impl better suggestion for `E0532`

Ezrashaw opened the pull request:
Fixes #106862

No test as there is already a test which is nearly identical to the example in the linked issue.

Ezrashaw committed:
error-msg: impl better suggestion for `E0532`

WaffleLapkin commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -1298,21 +1298,38 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
+                            Some(
+                                args.iter()
+                                    .map(|a| {
+                                        self.r.tcx.sess.source_map().span_to_snippet(*a).unwrap()
```
Adding a `span_to_snippet` to diagnostic logic doesn't sound like a good idea. Can you instead either
1. Leave field spans in place and add code around them (I think I'd prefer this), or
2. Save field names before hand

?

Ezrashaw replied:
Sorry, maybe I'm missing something but I thought the primary use of `SourceMap::span_to_snippet` was for diagnostic purposes.

I don't think we have access to the old fields, and even if we did that would require stringifying them which `span_to_snippet` already does.

WaffleLapkin replied:
Yes, `SourceMap::span_to_snippet` is intended for diagnostics, but we generally try to not use it, leaving it as a last resort when you can't implement the diagnostic otherwise :)

> EDIT: A mere 50 lines above is another span_to_snippet used for error messages.

Yes, but as I said, we'd like to remove most of them, not adding them is still encouraged.

----

You can use [`multipart_suggestion`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic/struct.Diagnostic.html#method.multipart_suggestion) to add field aliases without replacing the whole pattern. I.e. instead of

```
- S(a)
+ S { a: x }
```
you would do
```
S(a)
 - -

S { a: x }
 +++ +++++
```

Spans for adding aliases are just `arg_span.shrink_to_hi()`, `{}` are a bit harder, but something like `path_span.between(first_arg_span)` and `last_arg_span.between(the_whole_thing_span.shrink_to_hi())` should probably work.

Although maybe this will need a slight refactor of the surrounding code, if it's too hard leave a `// FIXME` instead and I'll fix it after this PR is merged.

Ezrashaw commented:
@WaffleLapkin I've added a test with all your cases. One of the suggestions was poor (`Foo { x: x }`), it's been updated.

WaffleLapkin commented:
Hm, I thought about another test case: what about the case, when the pattern and the struct have different number of fields? i.e. `struct S { x: u32, y: u32 }`/`S(a)`; `struct S { field: u32 }`/`S(x, y)`; `struct S {}`/`S(a)`; and `struct S { field: u32 }`/`S()`.

If the struct has more fields than the pattern, we could either suggest `..` or `excess_field0: _, excess_field1: _` (not sure what's better).

Also, is this code run on expressions too? If so, we should test it too (although it seems like expressions go in `PathSource::Expr`?... it's unclear from the docs and hard to guess while reviewing on gh)

bors commented:
:umbrella: The latest upstream changes (presumably #108992) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Ezrashaw commented:
Implemented "when the tuple is missing fields". Also rebased and updated tests.

I think that removing the `span_to_snippet` will be difficult because you have to compare the old and new field; you can't just suggest around the fields.

WaffleLapkin commented on tests/ui/pattern/issue-106862.stderr:
```diff
@@ -0,0 +1,48 @@
+LL |     FooB { x: i32, y: i32 }
+   |     ----------------------- `FooB` defined here
+...
+LL |         FooB() => (),
+   |         ^^^^^^ help: use struct pattern syntax instead: `FooB {  }`
```
> when the tuple is missing fields (eg `struct S { x: u32, y: u32 }`/`S(a)`) is to ignore extra struct fields, obviously this is bad. I'd suggest just putting them in as `S { field1: x, field2, field3 }`. People might want to use those fields and IMHO we shouldn't preemptively ignore them; worst case is a unused warning.

That sounds reasonable to me, can you implement it? Also, when you do, please add a `// run-rustfix` to the header of the file, so we know that after suggestions are applied the code compiles.

WaffleLapkin commented:
> I think that removing the `span_to_snippet` will be difficult because you have to compare the old and new field; you can't just suggest around the fields.

Alright, I see. I won't block this PR on removing `span_to_snippet` then.

Ezrashaw commented:
@WaffleLapkin Whoops, thought I had already done that but however.

@rustbot ready

WaffleLapkin commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -1289,25 +1289,43 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
+                            "",
+                            "pattern",
+                            Applicability::MachineApplicable,
+                            Some(
+                                args.iter()
+                                    .map(|a| {
+                                        self.r.tcx.sess.source_map().span_to_snippet(*a).unwrap()
+                                    })
+                                    .collect::<Vec<String>>(),
+                            ),
```
Last thing: can you remove the unwrap, falling back to the bad suggestion in case sources are not available?

Something like `args.iter().map(|&a| self.r.tcx.sess.source_map().span_to_snipper(a).ok()).collect::<Option<Vec<String>>>()` should work.

Ezrashaw commented:
@WaffleLapkin Done. Thanks for all the feedback.

WaffleLapkin commented:
@bors r+ rollup

bors commented:
:pushpin: Commit bd1732240b586733d8d9fb47cf9ef4cfc299c22e has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin bd1732240b586733d8d9fb47cf9ef4cfc299c22e -->
<!-- homu: {"type":"Approved","sha":"bd1732240b586733d8d9fb47cf9ef4cfc299c22e","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
